packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,315 @@
+    function _setProxies() internal {
+        bytes memory code = vm.getDeployedCode("Proxy.sol:Proxy");
+        uint160 prefix = uint160(0x420) << 148;
+        for (uint256 i = 0; i < PROXY_COUNT; i++) {
+            address addr = address(prefix | uint160(i));
+            if (_notProxied(addr)) {
+                continue;
+            }
+
+            vm.etch(addr, code);
+            vm.store(addr, PROXY_ADMIN_ADDRESS, bytes32(uint256(uint160(Predeploys.PROXY_ADMIN))));
+
+            if (_isDefinedPredeploy(addr)) {
+                address implementation = _predeployToCodeNamespace(addr);
+                vm.store(addr, PROXY_IMPLEMENTATION_ADDRESS, bytes32(uint256(uint160(implementation))));
+            }
+        }
+    }
+
+    /// @notice LEGACY_ERC20_ETH is not being predeployed
+    function _setImplementations() internal {
+        _setLegacyMessagePasser();
+        _setDeployerWhitelist();
+        _setWETH9();
+        _setL2StandardBridge();
+        _setL2CrossDomainMessenger();
+        _setSequencerFeeVault();
+        _setOptimismMintableERC20Factory();
+        _setL1BlockNumber();
+        _setGasPriceOracle();
+        _setGovernanceToken();
+        _setL1Block();
+    }
+
+    function _setLegacyMessagePasser() internal {
+        vm.etch(
+            _predeployToCodeNamespace(Predeploys.LEGACY_MESSAGE_PASSER),
+            vm.getDeployedCode("LegacyMessagePasser.sol:LegacyMessagePasser")
+        );
+    }
+
+    function _setDeployerWhitelist() internal {
+        vm.etch(
+            _predeployToCodeNamespace(Predeploys.DEPLOYER_WHITELIST),
+            vm.getDeployedCode("DeployerWhitelist.sol:DeployerWhitelist")
+        );
+    }
+
+    /// @dev Contract metadata appended to deployed bytecode will differ
+    ///      from previous L2 genesis output
+    function _setWETH9() internal {
+        vm.etch(
+            Predeploys.WETH9,
+            vm.getDeployedCode("WETH9.sol:WETH9")
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // string public name
+            hex"0000000000000000000000000000000000000000000000000000000000000000",
+            // "Wrapped Ether"
+            hex"577261707065642045746865720000000000000000000000000000000000001a"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // string public symbol
+            hex"0000000000000000000000000000000000000000000000000000000000000001",
+            // "WETH"
+            hex"5745544800000000000000000000000000000000000000000000000000000008"
+        );
+        vm.store(
+            Predeploys.WETH9,
+            // uint8 public decimals
+            hex"0000000000000000000000000000000000000000000000000000000000000002",
+            // 18
+            hex"0000000000000000000000000000000000000000000000000000000000000012"
+        );
+    }
+
+    function _setL2StandardBridge() internal {
+        vm.etch(
+            Predeploys.L2_STANDARD_BRIDGE,
```

Comment:
this should be the code addr since its proxied

Reply:
Looking at the current L2 genesis, the storage slots are set at the proxy address:

```json
"4200000000000000000000000000000000000010": {
      "code": "0x...",
      "storage": {
        "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000001",
        "0x0000000000000000000000000000000000000000000000000000000000000003": "0x0000000000000000000000004200000000000000000000000000000000000007",
        "0x0000000000000000000000000000000000000000000000000000000000000004": "0x0000000000000000000000000c8b5822b6e02cda722174f19a1439a7495a3fa6",
        "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc": "0x000000000000000000000000c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d30010",
        "0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103": "0x0000000000000000000000004200000000000000000000000000000000000018"
      },
      "balance": "0x0"
    },
```

and not the implementation address:

```json
"c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d30010": {
      "code": "0x0...",
      "balance": "0x0"
    },
```

Reply:
What we need to do for each predeploy:
- Etch the code for the implementation at the code address
- Call initialize on the implementation with zero/dummy values
- Etch the code for the proxy at the proxy address
- Ensuring the proxy references the correct implementation (the `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` slot is the implementation address slot
- Call initialize on the proxy with real values 

Those last two bullets might be doable in a single step with `upgradeAndCall`

Reply:
@mds1 Updated as per your comment! Also added this comment:

```
/// @dev The general flow of adding a predeploy is:
///      1. _setProxies uses vm.etch to set the Proxy.sol deployed bytecode for proxy address 0x420...0 to 0x420...799.
///      Additionally, the PROXY_ADMIN_ADDRESS and PROXY_IMPLEMENTATION_ADDRESS storage slots are set for the proxy address.
///      2. vm.etch sets the deployed bytecode for each predeploy at the implementation address (i.e. code namespace)
///      3. The `initialize` method is called at the implementation address with zero/dummy vaules if it exists
///      4. The `initialize` method is called at the proxy address with actual vaules if it exists
///      5. A `require` check to verify the expected implementation address is set for the proxy
```